### ISSUE TITLE
Layout: make more map-grid draw coordinates settings data-defined or controlled by QGIS

### ISSUE THREAD

**Issue by andreasneumann** [qgis/QGIS#38676]
**Feature description.**
Depending on the map rotation angle (variable @map_rotation) we want to be able to control whether to "Show All | Show Latitude Only | Show Longitude Only | Disabled). Basically, the goal is to avoid labels at very flat angles in relation to their reference map edge. The second goal of this feature is to avoid the necessity for manual changes for draw coordinate settings when the map rotation changes.

![image](https://user-images.githubusercontent.com/884476/92726756-65e9e180-f36e-11ea-9aea-c57842544c83.png)

I see two options here:

1. QGIS would switch automatically for me, taking into account threshold values. It would probably mean a fifth option, something like "Automatic switch between lat/lon". E.g. when map_rotation is between >315 or <45 "Show Longitude", when >=45 and <= 135 switch to Show Latitude, and so on.
2. Make it data-defined and let the user control it by himself based on his own rules taking into account @map_rotation

For me, both options would be fine. 1. is more convenient, 2. more flexible

**Comment by andreasneumann** [comment]
@olivierdalang and @nyalldawson - what do you think about the two options I listed above. Which one should we implement? 1), 2) or both?

**Comment by olivierdalang** [comment]
@andreasneumann 
With a orthogonal grid (same CRS for the map and the grid), can't you achieve the desired effect by choosing "Show all" and setting a threshold angle to 45° ?

**Comment by andreasneumann** [comment]
@olivierdalang right - this is exactly what I want. Thanks for the hint!

There is one small problem though: when you exactly hit the threshold value (e.g. 45 deg, or 135 deg map rotation with 45 deg skip low angled annotations), then neither latitude nor longitude are drawn. Any idea how to solve this?

**Comment by olivierdalang** [comment]
@andreasneumann 
This case would require a fix in the code (put a `<=` instead of a `<`).

But then you'd end up with both latitudes and longitudes...

Though not very elegant, using 44.999° (or 45.001°) instead of an exact 45° should work without any visible difference ? That would avoid adding yet another option/setting in the already overcrowded (in part thanks to me :-) ) map grid settings widget... 

If that doesn't work for you, I'd go for option B/, as option A/ may not work well in some scenarios (e.g. when you have different CRS between the map and the grid, the grid may not be orthogonal, so that latitude/longitude may not all fall on the same map border). It would also avoid adding another option.

As a general thought, maybe at some point we could simplify the UX by having all options (except hide/show) apply to all edges (and thus removing many comboboxes). Users could just duplicate grids if they want different options on different edges. It would be a bit less intimidating, esp. since most of the time, you want the same settings on all edges.

**Comment by andreasneumann** [comment]
@olivierdalang: yes, I think setting the threshold to 44.999 is acceptable here.

However, when testing it, I saw that the input spin box for "Skip low angled annotations" only accepts full degrees ;-(

So we need to allow fractional values here ... and then we can close this ticket.

**Comment by olivierdalang** [comment]
@andreasneumann Actually I meant 44.99 for map rotation, not for the threshold, so that only one of latitude/longitude gets drawn per edge.

But indeed being able to input decimal values for the "Skip low angled annotations" parameters would be good, and indeed too that this parameter should (as said in the tooltip) skip only values **below** that angle.

Here's the PR : https://github.com/qgis/QGIS/pull/38702 (I leave it as draft for some time in case there's more polishing to be done before around the rotated tick feature before the release)

**Comment by andreasneumann** [comment]
@olivierdalang - I tested with the recent master today, and it doesn't work for me yet.

I set the "Skip low angle annotations" to 44.9 (I can now enter fractional values - so that part works):
![image](https://user-images.githubusercontent.com/884476/93196287-8210c880-f74a-11ea-8331-0b80d8ec4f67.png)

But if my rotation angle is at 45:
![image](https://user-images.githubusercontent.com/884476/93196346-96ed5c00-f74a-11ea-82b4-c7a4529844d4.png)

I still get both longitude and latitude labels:
![image](https://user-images.githubusercontent.com/884476/93196430-b71d1b00-f74a-11ea-9965-0c18d1da6c26.png)

Only if my map rotation angle changes to 44 or 46 or below/above, then it works fine.

**Comment by olivierdalang** [comment]
@andreasneumann 
Sorry Andreas, I meant 44.99° (or 50.01) for map rotation, and 45° for skip low angled ticks. This way you'd have only one of the two coordinates showing up on each side. Does that work for you ?

**Comment by andreasneumann** [comment]
> @andreasneumann
> Sorry Andreas, I meant 44.99° (or 50.01) for map rotation, and 45° for skip low angled ticks. This way you'd have only one of the two coordinates showing up on each side. Does that work for you ?

Yes - I know. However, this is not very user friendly. Imagine the case where there is a large organization with centralized map templates. It is acceptable for the author / admin of the project to work with such fractional numbers in the map grid settings, but rotating a map would be done by any inexperienced user. Then I'd have to inform all of the occasional QGIS users in this organization that 45, 135, 225 and 315 are all "forbidden" numbers for map rotation? QGIS does, by default 1 degree steps in the spin box of the map rotation (makes sense) and eventually, people will end up at these "forbidden" map rotation values. Then the local QGIS support staff would get telephones - why is the map grid labeling disappearing?

Can we try to find a better solution? I am not so satisfied with this proposal.

**Comment by olivierdalang** [comment]
Sure, I agree it's not very elegant...

Let's do solution 2. (making the dropdown data-definable) then.

**Comment by olivierdalang** [comment]
@andreasneumann 
Here's the PR, please have a look to see if it's ok for you : https://github.com/qgis/QGIS/pull/39103
